test/opt/test_eliminate_dead_const.cpp

```diff
@@ -0,0 +1,717 @@
+            {
+                "%used_int = OpConstant %int 1",
+                "%used_double = OpConstant %double 3.14159265358979",
+            },
+            /*  .main_insts = */
+            {
+                "%int_var = OpVariable %_pf_int Function",
+                "OpStore %int_var %used_int",
+                "%double_var = OpVariable %_pf_double Function",
+                "OpStore %double_var %used_double",
+            },
+            /* .dead_consts = */
+            {
+                "%dead_bool = OpConstantTrue %bool",
+                "%dead_float = OpConstant %float 2.5",
+                "%dead_inner_struct = OpConstantComposite %inner_struct %dead_bool %used_int %dead_float %used_double",
+                "%dead_int = OpConstant %int 2",
+                "%dead_double = OpConstant %double 1.428571428514",
+                "%dead_outer_struct = OpConstantComposite %outer_struct %dead_inner_struct %dead_int %dead_double",
+            },
+        },
+        // clang-format on
+    })));
+
+INSTANTIATE_TEST_CASE_P(
+    ScalarTypeSpecConstants, EliminateDeadConstantTest,
+    ::testing::ValuesIn(std::vector<EliminateDeadConstantTestCase>({
+        // clang-format off
+        // All scalar type spec constants.
+        {
+            /* .used_consts = */
+            {
+                "%used_bool = OpSpecConstantTrue %bool",
+                "%used_uint = OpSpecConstant %uint 2",
+                "%used_int = OpSpecConstant %int 2",
+                "%used_float = OpSpecConstant %float 2.5",
+                "%used_double = OpSpecConstant %double 1.428571428514",
+            },
+            /*  .main_insts = */
+            {
+                "%bool_var = OpVariable %_pf_bool Function",
+                "%uint_var = OpVariable %_pf_uint Function",
+                "%int_var = OpVariable %_pf_int Function",
+                "%float_var = OpVariable %_pf_float Function",
+                "%double_var = OpVariable %_pf_double Function",
+                "OpStore %bool_var %used_bool", "OpStore %uint_var %used_uint",
+                "OpStore %int_var %used_int", "OpStore %float_var %used_float",
+                "OpStore %double_var %used_double",
+            },
+            /* .dead_consts = */
+            {
+                "%dead_bool = OpSpecConstantTrue %bool",
+                "%dead_uint = OpSpecConstant %uint 2",
+                "%dead_int = OpSpecConstant %int 2",
+                "%dead_float = OpSpecConstant %float 2.5",
+                "%dead_double = OpSpecConstant %double 1.428571428514",
+            },
+        },
+        // clang-format on
+    })));
+
+INSTANTIATE_TEST_CASE_P(
+    VectorTypeSpecConstants, EliminateDeadConstantTest,
+    ::testing::ValuesIn(std::vector<EliminateDeadConstantTestCase>({
+        // clang-format off
+        // Bool vector type spec constants. One vector has all component dead,
+        // another vector has one dead boolean and one used boolean.
+        {
```

Comment:
I feel like it may be ok to just do spotcheck instead of enumerating all possible types here. But I don't care that much.